src/servers/http_server_v2.cc

```diff
@@ -1096,41 +1099,109 @@ HTTPAPIServerV2::HandleServerMetadata(evhtp_request_t* req)
+        *v_idx += 1;
+      }
+
+      memcpy(json_base + offset, base, base_size);
+      offset += base_size;
+    }
+  }
+
+  if (length_tmp != 0) {
+    return TRTSERVER_ErrorNew(
+        TRTSERVER_ERROR_INVALID_ARG,
+        std::string(
+            "unexpected size for request JSON, expecting " +
+            std::to_string(length_tmp) + " more bytes")
+            .c_str());
+  }
+
+  document->Parse(json_base, length);
+  if (document->HasParseError()) {
+    return TRTSERVER_ErrorNew(
+        TRTSERVER_ERROR_INVALID_ARG,
+        std::string(
+            "failed to parse the request JSON buffer: " +
+            std::string(GetParseError_En(document->GetParseError())) + " at " +
+            std::to_string(document->GetErrorOffset()))
+            .c_str());
+  }
+
+  return nullptr;
+}
+
 TRTSERVER_Error*
 HTTPAPIServerV2::EVBufferToInput(
     const std::string& model_name, const InferRequestHeader& request_header,
     evbuffer* input_buffer,
     TRTSERVER_InferenceRequestProvider* request_provider,
-    AllocPayload* alloc_payload)
+    AllocPayload* alloc_payload, size_t header_length)
 {
   // Extract individual input data from HTTP body and register in
   // 'request_provider'. The input data from HTTP body is not
-  // necessarily contiguous so need to copy into a contiguous
-  // buffer to be able to parse it into a json.
+  // necessarily contiguous so may need to register multiple input
+  // "blocks" for a given input.
   //
-  // Get the addr and size of from the evbuffer.
-  int buffer_len = evbuffer_get_length(input_buffer);
-  char json_buffer[buffer_len];
-  ev_ssize_t extracted_size =
-      evbuffer_copyout(input_buffer, &json_buffer, buffer_len);
-  if (extracted_size != buffer_len) {
-    return TRTSERVER_ErrorNew(
-        TRTSERVER_ERROR_INVALID_ARG,
-        "failed to parse request buffer into json");
+  // Get the addr and size of each chunk of input data from the
+  // evbuffer.
+  struct evbuffer_iovec* v = nullptr;
+  int v_idx = 0;
+
+  int n = evbuffer_peek(input_buffer, -1, NULL, NULL, 0);
+  if (n > 0) {
+    v = static_cast<struct evbuffer_iovec*>(
+        alloca(sizeof(struct evbuffer_iovec) * n));
+    if (evbuffer_peek(input_buffer, -1, NULL, v, n) != n) {
+      return TRTSERVER_ErrorNew(
+          TRTSERVER_ERROR_INTERNAL, "unexpected error getting input buffers");
+    }
   }
 
+  // Extract just the json header from the complete buffer
   rapidjson::Document document;
   rapidjson::Document::AllocatorType& allocator = document.GetAllocator();
-  document.Parse(json_buffer);
-  if (document.HasParseError()) {
-    return TRTSERVER_ErrorNew(
-        TRTSERVER_ERROR_INVALID_ARG,
-        std::string(
-            "failed to parse the request buffer: " +
-            std::string(GetParseError_En(document.GetParseError())) + " at " +
-            std::to_string(document.GetErrorOffset()))
-            .c_str());
+  int buffer_len = 0;
+  if (header_length == 0) {
```

Comment:
Then clause doesn't do anything... and also if the kInferHeaderContentLengthHTTPHeader is not given then you don't read any json?

Reply:
Fixed this check